lib/fog/compute/xen_server/real.rb

```diff
@@ -14,7 +14,14 @@ def initialize(options={})
           @port        = options[:xenserver_port] || 80
           @verify_mode = options[:xenserver_verify_mode] || OpenSSL::SSL::VERIFY_PEER
           @connection  = Fog::XenServer::Connection.new(@host, @port, @use_ssl, @verify_mode, @timeout)
+
           @connection.authenticate(@username, @password)
+
+          @connection.master_slave_request
+          if @connection.slave?
+            @connection  = Fog::XenServer::Connection.new(@connection.master, @port, @use_ssl, @verify_mode, @timeout)
+            @connection.authenticate(@username, @password)
+          end
```

Comment:
Wayyy better. =)
Maybe we could make the `#master_slave_request` private and call it from inside `#slave?` and `#master`?